### PR TITLE
[Routing] Miscellanious

### PR DESCRIPTION
**Pull request by ghost**
I closed #41222 because i had too many conflicts on the branch, my bad.

| Q             | A
| ------------- | ---
| Branch?       | 5.x 
| Bug fix?      | no
| New feature?  | no 
| Deprecations? | no
| License       | MIT

Hi, I've been looking at the Symfony Routing source code quite a bit in order to write about it. I'm doing a little PR because I've adapted a few things for my needs and it doesn't cost me anything to offer them to you. If you're interested in small changes like this, I'll be happy to help you and work on small non-priority tasks like typo, etc.

### RouteCompiler
 I find that `RouteCompiler::compilePattern` is more readable if you destructure `$match` inside the foreach. Also I would replace `$pos` with `$cursor` in the sense that `$pos` moves through `$pattern`. This part of the component is quite complex and I find that these arrangements helped me to better visualise the process. Considering that a first reading of the method requires going back and forth in the code and probably consulting external resources, futures contributors will be more comfortable with this rearrangement imho.

Before:
```php
foreach ($matches as $match) {
    $important = $match[1][1] >= 0;
    $varName = $match[2][0];
    // get all static text preceding the current variable
    $precedingText = substr($pattern, $pos, $match[0][1] - $pos);
    $pos = $match[0][1] + \strlen($match[0][0]);
```

After:
```php
foreach ($matches as $match) {
    [[$param, $paramPos], $importance, [$varName]] = $match;
    $important = $importance[1] >= 0;
    // get all static text preceding the current variable
    $precedingText = substr($pattern, $cursor, $paramPos - $cursor);
    $cursor = $paramPos + \strlen($param);
```

### UrlMatcher
I don't know if it's deliberate but `->setContext` and `->matchRequest` are not tested in `UrlMatcherTest`. I've added two tests if you want.

### RouteTest
I think it's a good idea to create a specific method to test that fluent interfaces are respected. It avoids forgetting things like in `RouteTest` and it has already been implemented in `RequestContextTest`.

### CODE REVIEW

**Comment by marcel-jonquieres** [comment]
This test does not make sens: https://github.com/symfony/symfony/blob/8ed996426d549e8789c304feb52dd770669fca5c/src/Symfony/Component/Routing/Tests/Matcher/UrlMatcherTest.php#L912

File: src/Symfony/Component/Routing/Tests/Fixtures/Matcher/CustomUrlMatcher.php [hunk 1]
```diff
@@ -0,0 +1,13 @@
+<?php
+
+namespace Symfony\Component\Routing\Tests\Fixtures\Matcher;
+
+use Symfony\Component\Routing\Matcher\UrlMatcher;
+
+class CustomUrlMatcher extends UrlMatcher
```
Comment (derrabus) [inline src/Symfony/Component/Routing/Tests/Fixtures/Matcher/CustomUrlMatcher.php]: What's the purpose of this fixture?

File: src/Symfony/Component/Routing/Tests/Fixtures/ServiceProvider/ExpressionLanguageServiceProvider.php [hunk 2]
```diff
@@ -0,0 +1,18 @@
+<?php
+
+namespace Symfony\Component\Routing\Tests\Fixtures\ServiceProvider;
+
+use Symfony\Component\Routing\RouterInterface;
+use Symfony\Contracts\Service\ServiceProviderInterface;
+
+class ExpressionLanguageServiceProvider implements ServiceProviderInterface
```
Comment (derrabus) [inline src/Symfony/Component/Routing/Tests/Fixtures/ServiceProvider/ExpressionLanguageServiceProvider.php]: What's the purpose of this fixture?

File: src/Symfony/Component/Routing/Tests/RouteTest.php [hunk 3]
```diff
@@ -67,19 +66,16 @@ public function testOptions()
...
         $this->assertEquals(['foo' => 'foo', 'bar' => 'bar', 'compiler_class' => 'Symfony\\Component\\Routing\\RouteCompiler'], $route->getOptions(), '->addDefaults() keep previous defaults');
     }
 
     public function testOption()
     {
-        $route = new Route('/{foo}');
-        $this->assertFalse($route->hasOption('foo'), '->hasOption() return false if option is not set');
-        $this->assertEquals($route, $route->setOption('foo', 'bar'), '->setOption() implements a fluent interface');
+        $route = new Route('/{foo}', [], [], ['foo' => 'bar']);
+        $this->assertFalse($route->hasOption('baz'), '->hasOption() return false if option is not set');
```
Comment (derrabus) [inline src/Symfony/Component/Routing/Tests/RouteTest.php]: You are changing the meaning of the test here. The test made sure that an option that did not exist before the call to `setOption()` is properly set afterwards. You have removed the call to `setOption()` so we're not testing anymore that `setOption()` actually has an effect other than returning the `$route` object itself.

File: src/Symfony/Component/Routing/Tests/Matcher/UrlMatcherTest.php [hunk 4]
```diff
@@ -22,6 +23,24 @@
...
 {
+    public function testContext()
+    {
+        $matcher = $this->getUrlMatcher(new RouteCollection());
+
+        $matcher->setContext($requestContext = new RequestContext('bar'));
+        $this->assertSame($requestContext, $matcher->getContext());
+    }
+
+    public function testMatchRequest()
```
Comment (derrabus) [inline src/Symfony/Component/Routing/Tests/Matcher/UrlMatcherTest.php:26]: Where's the additional value in this test? Doesn't `testPatternMatchAndParameterReturn()` already cover everything that is tested here?

**Review by derrabus — CHANGES_REQUESTED** [review]
Hello and thank you for your first contribution.

* We are usually very conservative about code style changes like the ones you applied to `RouteCompiler`. If you want to contribute to Symfony, this kind of changes is not the best way to start. Readability is a highly subjective matter and PRs that don't fix an actual problem are very likely to be rejected.

* A pull request should have a meaningful title. This makes it easier for other contributors to decide whether this is a topic that they can perform a code review for or help with input. Furthermore, we generate the ChangeLog from the PR titles. "Miscellanious" does not really fulfill that requirement. If you are addressing multiple issues with your PR, it should probably be split.

* New tests are always welcome if they add value and don't duplicate existing tests.

* Unless you develop a new feature, your changes should be applied to the oldest maintained branch where they apply. In your case, this would probably be 4.4.

You can find more details on how we work with this repository in our documentation:

https://symfony.com/doc/current/contributing/index.html

File: src/Symfony/Component/Routing/Tests/Fixtures/ServiceProvider/ExpressionLanguageServiceProvider.php (hunk 2 above)
Comment (ghost) [reply src/Symfony/Component/Routing/Tests/Fixtures/ServiceProvider/ExpressionLanguageServiceProvider.php]: Its' an error, i probably add this by trying to rebase. It was deleted.

File: src/Symfony/Component/Routing/Tests/Fixtures/Matcher/CustomUrlMatcher.php (hunk 1 above)
Comment (ghost) [reply src/Symfony/Component/Routing/Tests/Fixtures/Matcher/CustomUrlMatcher.php]: Its' an error, i probably add this by trying to rebase. It was deleted.

**Comment by ghost** [comment]
@derrabus 
I understand your *conservatism*, I suspected it but it didn't cost me anything to suggest you.

Btw, if you have some idea of how I can contribute on new features, I might be interested, thanks for your time.

**Comment by derrabus** [comment]
Symfony lives from contributions from its community. I appreciate it that you want to give it a try. Here's what you can do.

* You can also browse open pull requests and see if you can help reviewing, testing or completing them.

* Most contribution come from developers that actively use Symfony in their projects and hit a bug or missing a feature that could be useful for them.

* Another way to contribute is to test your own projects with dev snapshots of Symfony and report or even fix issues that you find.

* Finally, there's a "Help Wanted" label on our issue tracker. You might want to browse issues with that label to find some inspiration: https://github.com/symfony/symfony/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+wanted%22

**Pull request closed without merging**